File: 配套资源/配套代码/程序设计竞赛专题挑战教程-代码/第7章小明的背包2.java
//小明的背包2  lanqiaoOJ 1175

import java.util.Scanner;

public class Main {
        public static void main(String[] args) {
            Scanner input = new Scanner(System.in);
            int n=input.nextInt();
            int m=input.nextInt();
            int w[]=new int [n+1];
            int v[]=new int [n+1];
            int dp[]=new int [m+1];
            for (int i=1;i<=n;i++)
            {
                w[i]=input.nextInt();
                v[i]=input.nextInt();
            }
            for (int i=1;i<=n;i++)
                for (int j=w[i];j<=m;j++)
                {
                    dp[j]=max(dp[j],dp[j-w[i]]+v[i]);
                }
            System.out.println(dp[m]);


        }
        static int max(int a,int b) {return a >=b ? a:b;}
 }

File: 配套资源/配套代码/程序设计竞赛专题挑战教程-代码/第7章小明的背包2.py
#小明的背包2  lanqiaoOJ 1175

n, C = map(int, input().split())
dp = [0]*(C+1)
for i in range(n):
    ci, wi = map(int, input().split())
    for j in range(ci, C+1):
        dp[j] +=  dp[j-ci]+wi
print(dp[-1])
